core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java

```diff
@@ -398,25 +428,57 @@ public synchronized void write(byte[] bs, int off, int len)
 	}
 
 	// Implement timeout with a scheduled task
-	ScheduledFuture<Object> sf = null;
+	ScheduledFuture<String> sf = null;
 	try {
-	    try {
-		if (timeout > 0)
-		    sf = ses.schedule(timeoutTask,
-					timeout, TimeUnit.MILLISECONDS);
-	    } catch (RejectedExecutionException ex) {
-		// ignore it; Executor was shut down by another thread,
-		// the following write should fail with IOException
-	    }
-	    os.write(bs, off, len);
+        try {
+            if (timeout > 0)
+                sf = ses.schedule(timeoutTask,
+                    timeout, TimeUnit.MILLISECONDS);
+        } catch (RejectedExecutionException ex) {
+            if (!socket.isClosed()) {
+                throw new IOException("Write aborted due to timeout not enforced", ex);
+            }
+        }
+
+        try {
+            os.write(bs, off, len);
+        } catch (IOException e) {
+            e = sf == null ? e : new IOException(handleTimeoutTaskResult(sf), e);
+            throw e;
+        }
 	} finally {
 	    if (sf != null)
-		sf.cancel(true);
-	}
+		    sf.cancel(true);
+	    }
     }
 
     @Override
     public void close() throws IOException {
 	os.close();
```

Comment:
Think this should be:
 os.close();
  if (sf != null)
    sf.cancel(true);
 }

Thoughts?

Reply:
good point, looks like we should make sf a member variable so that it would be accessible by close method

Reply:
fixed